Add description about custom profile (#48154)
kubernetes/website: content/en/docs/tasks/debug/debug-application/debug-running-pod.md

```shell
kubectl delete pod myapp
```

### Custom Profile {#custom-profile}

{{< feature-state for_k8s_version="v1.31" state="beta" >}}

You can define a partial container spec as a custom profile in either YAML or JSON format.
The custom profile can be applied to an ephemeral container or a debugging container in a copied Pod 
or a debugging Pod using the `--custom` flag.

{{< note >}}
Custom profile only supports the modification of the ephemeral container or debugging container spec.
It does not support the modification of the Pod spec.
Modifications for `command`, `image`, `lifecycle`, `name` and `volumeDevices` fields 
via a custom profile are not allowed.
{{< /note >}}

Create a Pod named myapp as an example:

```shell
kubectl run myapp --image=busybox:1.28 --restart=Never -- sleep 1d
```

Create a custom profile in YAML or JSON format.
Here, create a YAML format file named `custom-profile.yaml`:

> {{% code_sample file="debug/custom-profile.yaml" %}}

Reviewer: Please consider embed this YAML directly because it is not a self-contained k8s resource manifest for users to try out.

Author: @tengqm 
OK, I've addressed.

```shell
kubectl delete pod myapp
```

> ### Custom Profile {#custom-profile}

Reviewer 2: ```suggestion
## Custom profiles for ephemeral containers {#custom-profile}
```

Author: Same as Debugging Profiles above, custom profile can be applied not only to an ephemeral containers, but also to a copied Pods or a debugging pods to debug nodes.
Although the following case uses an ephemeral container as an example, I think the title should remain more general, as it is now.
(While other examples can be provided, it would be redundant)

Furthermore, I think custom profile is an extension of debugging profile.
So it would be better for the Custom Profile section under Debugging Profiles to be at level `###` instead of `##`, or to add a new comprehensive section that consolidates the Debug Profile and Custom Profile.

{{< feature-state for_k8s_version="v1.31" state="beta" >}}

> You can define partial container spec as a custom profile in either YAML or JSON format, and apply it using the `--custom` flag.

Reviewer 2: ```suggestion
You can define partial container spec as a custom profile in either YAML or JSON format, and
use it to create an ephemeral container based on that profile.
```

Author: As [above](https://github.com/kubernetes/website/pull/48154#discussion_r1789653676), custom profile is not limited to the use case of an ephemeral container, I've expressed that more clearly.

{{< note >}}

> - Custom profile only supports the modification of the debug container spec. It does not support the modification of the Pod spec of the debug target.
> - Modifications via custom profile is not allowed for certain fields such as command, image, lifecycle, volume devices and container name. In the future, more fields can be added to the disallowed list if required.

Reviewer 2: How about:
```suggestion
Custom profiles only support modification of the ephemeral debug container. You cannot use
custom profiles to modify running containers.

Modifications via a custom profile are not allowed for certain fields, including: `command`, `image`, `lifecycle`, `name`, and fields that define access to storage.
```
?

Author: As [above](https://github.com/kubernetes/website/pull/48154#discussion_r1789653676), I think we should not limit to just an ephemeral container.
And last sentence in your suggestion is good for me, so I've addressed.

{{< /note >}}

First, create a Pod named myapp as an example:

```shell
kubectl run myapp --image=busybox:1.28 --restart=Never -- sleep 1d
```

Create a custom profile in a YAML file named `custom-profile.yaml`:

{{% code_sample file="debug/custom-profile.yaml" %}}

You can also use JSON format for the custom profile:

> {{% code_sample file="debug/custom-profile.json" %}}

Reviewer 2: For this, see if you can use [tabs](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs).

Author: Yes, that's good!
I've addressed.

Then, debug the Pod using an ephemeral container with the custom profile:

```shell

> kubectl debug -it myapp --image=busybox:1.28 --target=myapp --custom=custom-profile.yaml

Reviewer 3: Since the default is `legacy`, would it be better to pass a profile e.g. `--profile=general`?

Author: OK, I've addressed

```

You can check that the ephemeral container was created with the custom profile applied:

```shell

> kubectl get po myapp -o jsonpath='{.spec.ephemeralContainers[0].env}'

Reviewer 2: In docs, we avoid abbreviating too much:
```suggestion
kubectl get po myapp -o jsonpath='{.spec.ephemeralContainers[0].env}'
```

Author: mistake too...

```

```
[{"name":"ENV_VAR_1","value":"value_1"},{"name":"ENV_VAR_2","value":"value_2"}]
```

```shell

> kubectl get po myapp -o jsonpath='{.spec.ephemeralContainers[0].securityContext}'

Reviewer 2: In docs, we avoid abbreviating too much:
```suggestion
kubectl get pod myapp -o jsonpath='{.spec.ephemeralContainers[0].securityContext}'
```

Author: Sorry, it's my mistake...
I've addressed.

```shell
kubectl delete pod myapp
```

### Custom Profile {#custom-profile}

{{< feature-state for_k8s_version="v1.31" state="beta" >}}

You can define a partial container spec as a custom profile in either YAML or JSON format.
The custom profile can be applied to an ephemeral container or a debugging container in a copied Pod 
or a debugging Pod using the `--custom` flag.

{{< note >}}
Custom profile only supports the modification of the ephemeral container or debugging container spec.
It does not support the modification of the Pod spec.
Modifications for `command`, `image`, `lifecycle`, `name` and `volumeDevices` fields 

> via a custom profile are not allowed.

Reviewer: This paragraph doesn't make sense.
If modifications to a Pod spec is not supported, why are we talking about the specific fields then?

```shell
kubectl delete pod myapp
```

### Custom Profile {#custom-profile}

{{< feature-state for_k8s_version="v1.31" state="beta" >}}

> You can define a partial container spec as a custom profile in either YAML or JSON format, and apply it to an ephemeral container or a debugging container in a copied Pod or a debugging Pod using the `--custom` flag.

Reviewer: A long sentence like this is difficult to parse. Please consider a revision.
Also, please try manually wrap long lines as we do for other paragraphs.

{{< note >}}
Custom profile only supports the modification of the debugging container spec. It does not support the modification of the Pod spec of the debug target.

> Modifications via a custom profile are not allowed for certain fields, including: `command`, `image`, `lifecycle`, `name`, and fields that define access to storage.

Reviewer: "certain fields" are vague. Please be more specific.

Author: I've indicated more specifically in line with the [implementation](https://github.com/kubernetes/kubernetes/blob/0b91def6bcbdde82206ec779791762fb6bc337af/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go#L396-L400).

{{< /note >}}

First, create a Pod named myapp as an example:

```shell
kubectl run myapp --image=busybox:1.28 --restart=Never -- sleep 1d
```

Create a custom profile in YAML or JSON format.
Here, create a YAML format file named `custom-profile.yaml`:

{{< tabs name="custom_profiles" >}}

> {{< tab name="YAML" codelang="yaml" >}}

Reviewer: Since the YAML and JSON files are equivalent, we don't need to show both.
The maintenance burden overweighs the convenience brought to users.

Author: I've fixed to show only YAML format file.

env:
- name: ENV_VAR_1
  value: value_1
- name: ENV_VAR_2
  value: value_2
securityContext:
  capabilities:
    add:
    - NET_ADMIN
    - SYS_TIME

{{< /tab >}}
{{< tab name="JSON" codelang="json" >}}
{
  "env": [
    {
      "name": "ENV_VAR_1",
      "value": "value_1"
    },
    {
      "name": "ENV_VAR_2",
      "value": "value_2"
    }
  ],
  "securityContext": {
    "capabilities": {
      "add": [
        "NET_ADMIN",
        "SYS_TIME"
      ]
    }
  }
}
{{< /tab >}}
{{< /tabs >}}

> Then, debug the Pod using an ephemeral container with the custom profile:

Reviewer: Instead of "first", "then", please try read some existing task pages and see how a task page is supposed to be structured.
